Swift: InvocationTargetException due to cleanup-*/database-* files

Downloading data from the repo using the swift plugin fails with an exception (`sy down` and `sy ls-remote` commands for example).

As an example, with `sy down` I get the following:
```
21-7-15 23:44:22.566 | RetriableFeatur | main           | WARN : Transfer method failed. 1/3 retries. Sleeping 3000ms ...
org.syncany.plugins.transfer.StorageException: java.lang.reflect.InvocationTargetException
        at org.syncany.plugins.transfer.files.RemoteFile.createRemoteFile(RemoteFile.java:130)
        at org.syncany.plugins.swift.SwiftTransferManager.list(SwiftTransferManager.java:208)
        at org.syncany.plugins.transfer.features.RetriableFeatureTransferManager$8.execute(RetriableFeatureTransferManager.java:146)
        at org.syncany.plugins.transfer.features.RetriableFeatureTransferManager.retryMethod(RetriableFeatureTransferManager.java:210)
        at org.syncany.plugins.transfer.features.RetriableFeatureTransferManager.list(RetriableFeatureTransferManager.java:143)
        at org.syncany.plugins.transfer.features.TransactionAwareFeatureTransferManager.list(TransactionAwareFeatureTransferManager.java:148)
        at org.syncany.operations.down.DownOperation.getRemoteCleanupFiles(DownOperation.java:695)
        at org.syncany.operations.down.DownOperation.execute(DownOperation.java:156)
        at org.syncany.cli.DownCommand.execute(DownCommand.java:56)
        at org.syncany.cli.CommandLineClient.runLocally(CommandLineClient.java:392)
        at org.syncany.cli.CommandLineClient.runCommand(CommandLineClient.java:382)
        at org.syncany.cli.CommandLineClient.start(CommandLineClient.java:198)
        at org.syncany.Syncany.main(Syncany.java:52)
Caused by: java.lang.reflect.InvocationTargetException
        at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
        at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:57)
        at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
        at java.lang.reflect.Constructor.newInstance(Constructor.java:526)
        at org.syncany.plugins.transfer.files.RemoteFile.createRemoteFile(RemoteFile.java:127)
        ... 12 more
Caused by: org.syncany.plugins.transfer.StorageException: database-LZeYqizikbLuMEnWGPBK-0000000004: remote filename pattern does not match: cleanup-([0-9]+) expected.
        at org.syncany.plugins.transfer.files.CleanupRemoteFile.validateName(CleanupRemoteFile.java:64)
        at org.syncany.plugins.transfer.files.RemoteFile.<init>(RemoteFile.java:73)
        at org.syncany.plugins.transfer.files.CleanupRemoteFile.<init>(CleanupRemoteFile.java:46)
        ... 17 more
```

The issue seems to be that in [SwiftTransferManager.java:228](https://github.com/syncany/syncany-plugin-swift/blob/develop/src/main/java/org/syncany/plugins/swift/SwiftTransferManager.java#L228) `SwiftTransferManager.getRemoteFilePath()` returns the same `/databases` path for both `cleanup-*` and `database-*` files. `SwiftTransferManager.list()` then gets confused at [SwiftTransferManager.java:212](https://github.com/syncany/syncany-plugin-swift/blob/develop/src/main/java/org/syncany/plugins/swift/SwiftTransferManager.java#L212) since it gets both `cleanup-*` and `database-*` files from the listing, but expects them to be of the specified type.

Versions:
 * Syncany: 0.4.6-alpha (from the deb package)
 * Swift plugin: 0.4.0-alpha from the repos + compiled the develop branch myself

--- a/build.gradle
+++ b/build.gradle
@@ -8,10 +8,10 @@ configurations {
 project.ext {
 	pluginId = "swift"
 	pluginName = "OpenStack Swift"
-	pluginVersion = "0.4.0-alpha"
-	pluginDebianVersion = "1"
+	pluginVersion = "0.4.1-alpha"
+	pluginDebianVersion = "1.1"
 	pluginDate = new Date()
-	pluginAppMinVersion = "0.4.0-alpha"
+	pluginAppMinVersion = "0.4.1-alpha"
 	pluginRelease = rootProject.ext.applicationRelease
 	pluginConflictsWith = ""
 

--- a/src/main/java/org/syncany/plugins/swift/SwiftTransferManager.java
+++ b/src/main/java/org/syncany/plugins/swift/SwiftTransferManager.java
@@ -83,14 +83,29 @@ public class SwiftTransferManager extends AbstractTransferManager {
 	public SwiftTransferManager(SwiftTransferSettings settings, Config config) {
 		super(settings, config);
 
+		boolean usingKeystone = false;
+		
 		AccountConfig authConfig = new AccountConfig();
 		authConfig.setUsername(settings.getUsername());
 		authConfig.setPassword(settings.getPassword());
-		authConfig.setTenantName(settings.getTenantName());
 		authConfig.setAuthUrl(settings.getAuthUrl());
+		
+		if(settings.getTenantName() != null && !settings.getTenantName().equals("")) {
+			authConfig.setTenantName(settings.getTenantName());
+			usingKeystone = true;
+		}
+		
+		if(settings.getTenantId() != null && !settings.getTenantId().equals("")) {
+			authConfig.setTenantId(settings.getTenantId());
+			usingKeystone = true;
+		}
 		
 		if(settings.getPreferredRegion() != null && !settings.getPreferredRegion().equals("")) {
 			authConfig.setPreferredRegion(settings.getPreferredRegion());
+		}
+		
+		if(usingKeystone) {
+			authConfig.setAuthenticationMethod(AuthenticationMethod.KEYSTONE);
 		}
 
 		this.account = new AccountFactory(authConfig).createAccount();

--- a/src/main/java/org/syncany/plugins/swift/SwiftTransferManager.java
+++ b/src/main/java/org/syncany/plugins/swift/SwiftTransferManager.java
@@ -96,11 +96,11 @@ public class SwiftTransferManager extends AbstractTransferManager {
 		this.account = new AccountFactory(authConfig).createAccount();
 
 		this.container = account.getContainer(settings.getContainer());
-		this.multichunksPath = "/multichunks";
-		this.databasesPath = "/databases";
-		this.actionsPath = "/actions";
-		this.transactionsPath = "/transactions";
-		this.tempPath = "/temporary";
+		this.multichunksPath = "multichunks";
+		this.databasesPath = "databases";
+		this.actionsPath = "actions";
+		this.transactionsPath = "transactions";
+		this.tempPath = "temp";
 	}
 
 	@Override

--- a/src/main/java/org/syncany/plugins/swift/SwiftTransferManager.java
+++ b/src/main/java/org/syncany/plugins/swift/SwiftTransferManager.java
@@ -33,6 +33,7 @@ import org.javaswift.joss.model.Account;
 import org.javaswift.joss.model.Container;
 import org.javaswift.joss.model.StoredObject;
 import org.syncany.config.Config;
+import org.syncany.plugins.swift.SwiftTransferSettings;
 import org.syncany.plugins.swift.SwiftTransferManager.SwiftReadAfterWriteConsistentFeatureExtension;
 import org.syncany.plugins.transfer.AbstractTransferManager;
 import org.syncany.plugins.transfer.StorageException;
@@ -85,8 +86,12 @@ public class SwiftTransferManager extends AbstractTransferManager {
 		AccountConfig authConfig = new AccountConfig();
 		authConfig.setUsername(settings.getUsername());
 		authConfig.setPassword(settings.getPassword());
+		authConfig.setTenantName(settings.getTenantName());
 		authConfig.setAuthUrl(settings.getAuthUrl());
-		authConfig.setAuthenticationMethod(AuthenticationMethod.BASIC);
+		
+		if(settings.getPreferredRegion() != null && !settings.getPreferredRegion().equals("")) {
+			authConfig.setPreferredRegion(settings.getPreferredRegion());
+		}
 
 		this.account = new AccountFactory(authConfig).createAccount();
 

--- a/src/main/java/org/syncany/plugins/swift/SwiftTransferManager.java
+++ b/src/main/java/org/syncany/plugins/swift/SwiftTransferManager.java
@@ -229,8 +229,13 @@ public class SwiftTransferManager extends AbstractTransferManager {
 		Map<String, T> remoteFiles = new HashMap<>();
 
 		for (StoredObject child : listing) {
-			T remoteFile = RemoteFile.createRemoteFile(child.getBareName(), remoteFileClass);
-			remoteFiles.put(child.getBareName(), remoteFile);
+			try {
+				T remoteFile = RemoteFile.createRemoteFile(child.getBareName(), remoteFileClass);
+				remoteFiles.put(child.getBareName(), remoteFile);
+			} catch(Exception e) {
+				logger.log(Level.INFO, "Cannot create instance of " + remoteFileClass.getSimpleName() + " for object " + child.getBareName()
+						+ "; maybe invalid file name pattern. Ignoring file.");
+			}
 		}
 
 		return remoteFiles;

--- a/src/main/java/org/syncany/plugins/swift/SwiftTransferSettings.java
+++ b/src/main/java/org/syncany/plugins/swift/SwiftTransferSettings.java
@@ -40,16 +40,20 @@ public class SwiftTransferSettings extends TransferSettings {
 	@Encrypted
 	public String password;
 	
-	@Element(name = "tenantName", required = true)
-	@Setup(order = 4, description = "Tenant name")
+	@Element(name = "tenantName", required = false)
+	@Setup(order = 4, description = "Tenant name (applicable if using Keystone to authenticate)")
 	public String tenantName;
 	
+	@Element(name = "tenantId", required = false)
+	@Setup(order = 5, description = "Tenant ID (applicable if using Keystone to authenticate)")
+	public String tenantId;
+	
 	@Element(name = "container", required = true)
-	@Setup(order = 5, description = "Target container to use")
+	@Setup(order = 6, description = "Target container to use")
 	public String container;
 	
 	@Element(name = "preferredRegion", required = false)
-	@Setup(order = 6, description = "Preferred Swift data region (leave blank to use first applicable region)")
+	@Setup(order = 7, description = "Preferred Swift data region (leave blank to use first applicable region)")
 	public String preferredRegion;
 
 	public String getAuthUrl() {
@@ -66,6 +70,10 @@ public class SwiftTransferSettings extends TransferSettings {
 	
 	public String getTenantName() {
 		return tenantName;
+	}
+	
+	public String getTenantId() {
+		return tenantId;
 	}
 
 	public String getContainer() {

--- a/src/main/java/org/syncany/plugins/swift/SwiftTransferSettings.java
+++ b/src/main/java/org/syncany/plugins/swift/SwiftTransferSettings.java
@@ -39,10 +39,18 @@ public class SwiftTransferSettings extends TransferSettings {
 	@Setup(order = 3, sensitive = true, description = "Password")
 	@Encrypted
 	public String password;
-
+	
+	@Element(name = "tenantName", required = true)
+	@Setup(order = 4, description = "Tenant name")
+	public String tenantName;
+	
 	@Element(name = "container", required = true)
-	@Setup(order = 4, description = "Target container to use")
+	@Setup(order = 5, description = "Target container to use")
 	public String container;
+	
+	@Element(name = "preferredRegion", required = false)
+	@Setup(order = 6, description = "Preferred Swift data region (leave blank to use first applicable region)")
+	public String preferredRegion;
 
 	public String getAuthUrl() {
 		return authUrl;
@@ -55,9 +63,17 @@ public class SwiftTransferSettings extends TransferSettings {
 	public String getPassword() {
 		return password;
 	}
+	
+	public String getTenantName() {
+		return tenantName;
+	}
 
 	public String getContainer() {
 		return container;
+	}
+	
+	public String getPreferredRegion() {
+		return preferredRegion;
 	}
 
 }

